Lib/test/test_importlib/test_util.py

```diff
@@ -758,6 +760,94 @@ def test_source_from_cache_path_like_arg(self):
         expect = os.path.join('foo', 'bar', 'baz', 'qux.py')
         self.assertEqual(self.util.source_from_cache(path), expect)
 
+    @contextlib.contextmanager
+    def bytecode_path(self, path):
+        # Patch directly in globals of the functions under test, because we have
+        # no direct access to the right _bootstrap_external module to patch
+
+        # cache_from_source and source_from_cache come from the same module,
+        # thus they share the same globals dict, so we only have to patch it
+        # once
+        _globals = self.util.cache_from_source.__globals__
+
+        _orig_path = _globals['bytecode_path']
+        _globals['bytecode_path'] = path
+        try:
+            yield
+        finally:
+            _globals['bytecode_path'] = _orig_path
+
+    @unittest.skipUnless(sys.implementation.cache_tag is not None,
+                         'requires sys.implementation.cache_tag not be None')
+    def test_cache_from_source_respects_PYTHONBYTECODEPATH(self):
+        # If PYTHONBYTECODEPATH env var is set, cache_from_source will return
+        # a bytecode path inside that directory (in a subdirectory mirroring the
+        # .py file's path) rather than in a __pycache__ dir next to the py file.
+        bytecode_paths = [
+            os.path.join(os.path.sep, 'tmp', 'bytecode'),
+            os.path.join(os.path.sep, 'tmp', '\u2603'),  # non-ASCII in path!
+            os.path.join(os.path.sep, 'tmp', 'trailing-slash') + os.path.sep,
+        ]
+        drive = os.path.splitdrive(os.getcwd())[0]
```

Comment:
Just for clarification, os.getcwd() always returns an absolute path?  or maybe that doesn't matter for splitdrive()?

Reply:
Python documentation isn't clear on the topic, but I checked the docs for `GetCurrentDirectory` on Windows (https://msdn.microsoft.com/en-us/library/windows/desktop/aa364934(v=vs.85).aspx) and libc `getcwd` (http://man7.org/linux/man-pages/man2/getcwd.2.html), which are how `os.getcwd` is implemented, and both are clearly defined to return absolute paths. And even more to the point, both the posix and Windows versions of `os.path.abspath` use "join with `os.getcwd()`" as their algorithm. Which is good, because even if I change the code here, in `_bootstrap_external.py` (where I don't have access to `os.path`) I'm also relying on "join with cwd" as a way to absolutize a path.